README.md
jQuery Scanner Detection
========================

jQuery Scanner Detection is a small plugin to detect when user use a scanner (barcode, QR Code...) instead of a keyboard, and call specific callbacks.


How use it
----------
To initialize the detection, 2 ways:

    $(selector).scannerDetection(); // Initialize with default options
    $(selector).scannerDetection({...}); // Initialize with an object that contains options
    $(selector).scannerDetection(function(){...}); // Initialize with a function that is onComplete callback

To simulate a scanning after initialisation:

    $(selector).scannerDetection('scanned string');

To disable the detection (deinitialize):

    $(selector).scannerDetection(false);

See a detailed tutorial with presets for different scanner models and first steps to take when starting [here](http://a.kabachnik.info/jquery-scannerdetection-tutorial.html).


Options
-------
### onComplete
Default: false
Callback after detection of a successful scanning
### onError
Default: false
Callback after detection of a unsuccessful scanning (scanned string in parameter)
### onReceive
Default: false
Callback after receiving and processing a char (scanned char in parameter)
### onKeyDetect
Default: false
Callback after detecting a keyDown (key char in parameter) - in contrast to onReceive, this fires for non-character keys like tab, arrows, etc. too!
### onScanButtonLongPressed
Default: false
Callback after detection of a successfull scan while the scan button was pressed and held down. This can only be used if the scan button behaves as a key itself (see scanButtonKeyCode). This long press event can be used to add a secondary action. For example, if the primary action is to count some items with barcodes (e.g. products at goods-in), it is comes very handy if a number pad pops up on the screen when the scan button is held. Large number can then be easily typed it instead of scanning fifty times in a row.
Note: this option requires scanButtonKeyCode to be set to a valid key code!
### timeBeforeScanTest
Default: 100
Wait duration (ms) after keypress event to check if scanning is finished
### avgTimeByChar
Default: 30
Average time (ms) between 2 chars. Used to do difference between keyboard typing and scanning
### minLength
Default: 6
Minimum length for a scanning
### endChar
Default: [9,13]
Chars to remove and means end of scanning
### startChar
Default: []
Chars to remove and means start of scanning
### ignoreIfFocusOn
Default: false
Ignore scans if the currently focused element matches this selector. Per example, if you set this option to 'input', scanner detection will be disable if an input is focused, when the scan occurs.
### scanButtonKeyCode
Default: false
Key code of the scanner hardware button (if the scanner button a acts as a key itself). Knowing this key code is important, because it is not part of the scanned code and must be ignored. On the other hand, knowing it can be usefull: pressing the button multiple times fast normally results just in one scan, but you still could count the number of times pressed, allowing the user to input quantities this way (typical use case would be counting product at goods-in).
### scanButtonLongPressThreshold
Default: 3
You can let the user perform some special action by pressing and holding the scan button. In this case the button will issue multiple keydown events. This parameter sets, how many sequential events should be interpreted as holding the button down.
### stopPropagation
Default: false
Stop immediate propagation on keypress event
### preventDefault
Default: false
Prevent default action on keypress event


Events
------
All callbacks are of type function and can also be bound as event listeners.
Like this, you can add multiple callbacks for each event.
Of course, events exist only if plugin is initialized on selector.

    $(selector)
        .bind('scannerDetectionComplete',function(e,data){...})
        .bind('scannerDetectionError',function(e,data){...})
        .bind('scannerDetectionReceive',function(e,data){...})

### scannerDetectionComplete
Callback after detection of a successful scanning
Event data: {string: "scanned string"}
### scannerDetectionError
Callback after detection of a unsuccessful scanning
Event data: {string: "scanned string"}
### scannerDetectionReceive
Callback after receive and process a char
Event data: {evt: {original keypress event}}
### scannerDetectionKeyDetect
Callback after detecting a keyDown - in contrast to onReceive, this fires for non-character keys like tab, arrows, etc. too!
Event data: {evt: {original keypress event}}

Requirements
------------
jQuery > 1.4.4. Note, that some odd side effects may occur before version 1.6, if your event handlers rely on using the event object (for example calling e.which from within the onKeyDetect handler).

Browser compatibility
---------------------
On old IE browser (IE<9), `Date.now()` and `Array.indexOf()` are not implemented.
If you plan to use this plugin on these browsers, please add jquery.scannerdetection.compatibility.js file before the plugin.


CHANGELOG.txt
1.2.1
  Add a onKeyDetect callback being called if any key (even non-chars) is detected
  Changed default scanButtonKeyCode to false (it turned out, that the previous value of '0' was actually used by some scanners)

1.2.0
  Add options startChar, ignoreIfFocusOn, scanButtonKeyCode, scanButtonLongPressThreshold

1.1.4
  Add script to extend compatibility to IE<9

1.1.3
  Change event.timestamp to Date.now() because Firefox has an issue with that

1.1.2
  Fix issue with tab key

1.1.1
  Fix issue with $(selector).scannerDetection('scanned string')

1.1.0
 Add events scannerDetectionComplete, scannerDetectionError, scannerDetectionReceive

1.0.0
 Initial version


jquery.scannerdetection.compatibility.js
/*
 * jQuery Scanner Detection
 *
 * Copyright (c) 2013 Julien Maurel
 *
 * Licensed under the MIT license:
 * http://www.opensource.org/licenses/mit-license.php
 *
 * Project home:
 * https://github.com/julien-maurel/jQuery-Scanner-Detection
 *
 * Version: 1.2.0
 *
 * Implement some functions that not exist on old IE browser (IE<9)
 */
if (!Date.now) {
  Date.now = function now() {
    return new Date().getTime();
  };
}

if (!Array.prototype.indexOf) {
  Array.prototype.indexOf = function (searchElement, fromIndex) {
    if (this === undefined || this === null) {
      throw new TypeError('"this" is null or not defined');
    }

    var length = this.length >>> 0; // Hack to convert object.length to a UInt32

    fromIndex = +fromIndex || 0;

    if (Math.abs(fromIndex) === Infinity) {
      fromIndex = 0;
    }

    if (fromIndex < 0) {
      fromIndex += length;
      if (fromIndex < 0) {
        fromIndex = 0;
      }
    }

    for (; fromIndex < length; fromIndex++) {
      if (this[fromIndex] === searchElement) {
        return fromIndex;
      }
    }

    return -1;
  };
}

jquery.scannerdetection.js
/*
 * jQuery Scanner Detection
 *
 * Copyright (c) 2013 Julien Maurel
 *
 * Licensed under the MIT license:
 * http://www.opensource.org/licenses/mit-license.php
 *
 * Project home:
 * https://github.com/julien-maurel/jQuery-Scanner-Detection
 *
 * Version: 1.2.1
 *
 */
(function ($) {
  $.fn.scannerDetection = function (options) {

    // If string given, call onComplete callback
    if (typeof options === "string") {
      this.each(function () {
        this.scannerDetectionTest(options);
      });
      return this;
    }

    // If false (boolean) given, deinitialize plugin
    if (options === false) {
      this.each(function () {
        this.scannerDetectionOff();
      });
      return this;
    }

    var defaults = {
      onComplete: false, // Callback after detection of a successfull scanning (scanned string in parameter)
      onError: false, // Callback after detection of a unsuccessfull scanning (scanned string in parameter)
      onReceive: false, // Callback after receiving and processing a char (scanned char in parameter)
      onKeyDetect: false, // Callback after detecting a keyDown (key char in parameter) - in contrast to onReceive, this fires for non-character keys like tab, arrows, etc. too!
      timeBeforeScanTest: 100, // Wait duration (ms) after keypress event to check if scanning is finished
      avgTimeByChar: 30, // Average time (ms) between 2 chars. Used to do difference between keyboard typing and scanning
      minLength: 6, // Minimum length for a scanning
      endChar: [9, 13], // Chars to remove and means end of scanning
      startChar: [], // Chars to remove and means start of scanning
      ignoreIfFocusOn: false, // do not handle scans if the currently focused element matches this selector
      scanButtonKeyCode: false, // Key code of the scanner hardware button (if the scanner button a acts as a key itself)
      scanButtonLongPressThreshold: 3, // How many times the hardware button should issue a pressed event before a barcode is read to detect a longpress
      onScanButtonLongPressed: false, // Callback after detection of a successfull scan while the scan button was pressed and held down
      stopPropagation: false, // Stop immediate propagation on keypress event
      preventDefault: false // Prevent default action on keypress event
    };

    if (typeof options === "function") {
      options = { onComplete: options }
    }
    if (typeof options !== "object") {
      options = $.extend({}, defaults);
    } else {
      options = $.extend({}, defaults, options);
    }

    this.each(function () {
      var self = this, $self = $(self), firstCharTime = 0, lastCharTime = 0, scannedString = '', isScanningComplete = false, testTimer = false, scanButtonCounter = 0;
      var initScannerDetection = function () {
        firstCharTime = 0;
        scannedString = '';
        scanButtonCounter = 0;
      };

      self.scannerDetectionOff = function () {
        $self.unbind('keydown.scannerDetection');
        $self.unbind('keypress.scannerDetection');
      }

      self.isFocusOnIgnoredElement = function () {
        if (!options.ignoreIfFocusOn) return false;
        if (typeof options.ignoreIfFocusOn === 'string') return $(':focus').is(options.ignoreIfFocusOn);
        if (typeof options.ignoreIfFocusOn === 'object' && options.ignoreIfFocusOn.length) {
          var focused = $(':focus');
          for (var i = 0; i < options.ignoreIfFocusOn.length; i++) {
            if (focused.is(options.ignoreIfFocusOn[i])) {
              return true;
            }
          }
        }
        return false;
      }

      self.scannerDetectionTest = function (s) {
        // If string is given, test it
        if (s) {
          firstCharTime = lastCharTime = 0;
          scannedString = s;
        }

        if (!scanButtonCounter) {
          scanButtonCounter = 1;
        }

        // If all condition are good (length, time...), call the callback and re-initialize the plugin for next scanning
        // Else, just re-initialize
        if (scannedString.length >= options.minLength && lastCharTime - firstCharTime < scannedString.length * options.avgTimeByChar) {
          if (options.onScanButtonLongPressed && scanButtonCounter > options.scanButtonLongPressThreshold) options.onScanButtonLongPressed.call(self, scannedString, scanButtonCounter);
          else if (options.onComplete) options.onComplete.call(self, scannedString, scanButtonCounter);
          $self.trigger('scannerDetectionComplete', { string: scannedString });
          initScannerDetection();
          return true;
        } else {
          if (options.onError) options.onError.call(self, scannedString);
          $self.trigger('scannerDetectionError', { string: scannedString });
          initScannerDetection();
          return false;
        }
      }

      $self.data('scannerDetection', { options: options })
        .unbind('.scannerDetection')
        .bind('keydown.scannerDetection', function (e) {
          // If it's just the button of the scanner, ignore it and wait for the real input
          if (options.scanButtonKeyCode !== false && e.which == options.scanButtonKeyCode) {
            scanButtonCounter++;

            // Cancel default
            e.preventDefault();
            e.stopImmediatePropagation();
          }
          // Add event on keydown because keypress is not triggered for non character keys (tab, up, down...)
          // So need that to check endChar and startChar (that is often tab or enter) and call keypress if necessary
          else if ((firstCharTime && options.endChar.indexOf(e.which) !== -1)
            || (!firstCharTime && options.startChar.indexOf(e.which) !== -1)) {
            // Clone event, set type and trigger it
            var e2 = jQuery.Event('keypress', e);
            e2.type = 'keypress.scannerDetection';
            $self.triggerHandler(e2);

            // Cancel default
            e.preventDefault();
            e.stopImmediatePropagation();
          }

          // Fire keyDetect event in any case!
          if (options.onKeyDetect) options.onKeyDetect.call(self, e);
          $self.trigger('scannerDetectionKeyDetect', { evt: e });

        }).bind('keypress.scannerDetection', function (e) {
          if (this.isFocusOnIgnoredElement()) return;
          if (options.stopPropagation) e.stopImmediatePropagation();
          if (options.preventDefault) e.preventDefault();

          if (firstCharTime && options.endChar.indexOf(e.which) !== -1) {
            e.preventDefault();
            e.stopImmediatePropagation();
            isScanningComplete = true;
          } else if (!firstCharTime && options.startChar.indexOf(e.which) !== -1) {
            e.preventDefault();
            e.stopImmediatePropagation();
            isScanningComplete = false;
          } else {
            if (typeof (e.which) != 'undefined') {
              scannedString += String.fromCharCode(e.which);
            }
            isScanningComplete = false;
          }

          if (!firstCharTime) {
            firstCharTime = Date.now();
          }
          lastCharTime = Date.now();

          if (testTimer) clearTimeout(testTimer);
          if (isScanningComplete) {
            self.scannerDetectionTest();
            testTimer = false;
          } else {
            testTimer = setTimeout(self.scannerDetectionTest, options.timeBeforeScanTest);
          }

          if (options.onReceive) options.onReceive.call(self, e);
          $self.trigger('scannerDetectionReceive', { evt: e });
        });
    });

    return this;
  };
})(jQuery);
